tidy3d/plugins/design/method.py

```diff
@@ -23,335 +20,780 @@ class Method(Tidy3dBaseModel, ABC):
-                task_name_mappings.append([task_name])
-            elif isinstance(sim, dict):
-                task_name_mappings.append({})
-                for name, _sim in sim.items():
-                    task_name = get_task_name(pt_index=1, sim_index=name, fn_kwargs=fn_kwargs)
-                    simulations[task_name] = _sim
-                    task_name_mappings[i][name] = task_name
-            else:
-                task_name_mappings.append([])
-                for j, _sim in enumerate(sim):
-                    task_name = get_task_name(pt_index=i, sim_index=j, fn_kwargs=fn_kwargs)
-                    simulations[task_name] = _sim
-                    task_name_mappings[i].append(task_name)
+        fn_args = self.sample(parameters)
+        for arg_dict in fn_args:
+            self._force_int(arg_dict, parameters)
+        return fn_args
 
-        # run in batch
-        batch_data = self._run_batch(simulations=simulations, path_dir=path_dir, **batch_kwargs)
-        task_id_dict = batch_data.task_ids
+    def _run(self, parameters: Tuple[ParameterType, ...], run_fn: Callable, console) -> Tuple[Any]:
+        """Defines the search algorithm."""
 
-        # run post processing on each data
-        result = []
-        task_ids = []
-        for task_names_i in task_name_mappings:
-            if isinstance(task_names_i, dict):
-                task_ids.append([task_id_dict[task_name] for task_name in task_names_i.values()])
-                kwargs_dict = {
-                    kwarg_name: batch_data[task_name]
-                    for kwarg_name, task_name in task_names_i.items()
-                }
-                val = fn_post(**kwargs_dict)
-            else:
-                task_ids.append([task_id_dict[task_name] for task_name in task_names_i])
-                args_list = (batch_data[task_name] for task_name in task_names_i)
-                val = fn_post(*args_list)
+        # get all function inputs
+        fn_args = self._assemble_args(parameters)
 
-            result.append(val)
+        # Run user function on sampled args
+        results = self._extract_output(run_fn(fn_args), sampler=True)
 
-        return fn_args, result, task_ids, batch_data
+        # None is aux_output and opt_output
+        return fn_args, results, None, None
 
 
-class MethodGrid(MethodIndependent):
+class MethodGrid(MethodSample):
     """Select parameters uniformly on a grid.
 
+    Size of the grid is specified by the parameter type,
+    either as the number of unique discrete values (ParameterInt, ParameterAny)
+    or with the num_points argument (ParameterFloat).
+
     Example
     -------
     >>> import tidy3d.plugins.design as tdd
     >>> method = tdd.MethodGrid()
     """
 
-    def sample(self, parameters: Tuple[ParameterType, ...], **kwargs) -> Dict[str, Any]:
-        """Defines how the design parameters are sampled on grid."""
+    def get_run_count(self, parameters: list) -> int:
```

Comment:
docstring

Reply:
Added all these. Does the docstring not inherit from the `@abstractmethod` defined in `Method`?

Reply:
that's a good question. I dont think it does unfortunately.

Reply:
That's good to know!

Reply:
the docstrings remain in subclasses as long as you dont overwrite the methods

```
>>> class A:
...     def method(self):
...             """docstring"""
...             pass
...
>>> class B(A):
...     pass
...
>>> B.method.__doc__
'docstring'
>>> class C(A):
...     def method(self):
...             pass
...
>>> C.method.__doc__
(nothing printed)
```